src/MobSpawner.cpp

```diff
@@ -340,6 +296,139 @@ bool cMobSpawner::CanSpawnHere(cChunk * a_Chunk, int a_RelX, int a_RelY, int a_R
 
 
 
+std::set<eMonsterType> cMobSpawner::GetAllowedMobTypes(EMCSBiome a_Biome)
+{
+	std::set<eMonsterType> ListOfSpawnables;
+	// Check biomes first to get a list of animals
+	switch (a_Biome)
+	{
+		// Nether mobs and endermen only - no other mobs in the nether
+		case biNether:
+		{
+			ListOfSpawnables.insert(mtGhast);
+			ListOfSpawnables.insert(mtMagmaCube);
+			ListOfSpawnables.insert(mtZombiePigman);
+			ListOfSpawnables.insert(mtEnderman);
+			return ListOfSpawnables;
+		}
+
+		// Endermen only - no other mobs in the end
+		case biEnd:
+		{
+			ListOfSpawnables.insert(mtEnderman);
+			return ListOfSpawnables;
+		}
+
+		// Mooshroom only - no other mobs on mushroom islands
+		case biMushroomIsland:
+		case biMushroomShore:
+		{
+			ListOfSpawnables.insert(mtMooshroom);
+			return ListOfSpawnables;
+		}
+
+		// Add Squid in ocean and river biomes
+		case biOcean:
+		case biFrozenOcean:
+		case biFrozenRiver:
+		case biRiver:
+		case biDeepOcean:
+		{
+			ListOfSpawnables.insert(mtSquid);
+			ListOfSpawnables.insert(mtGuardian);
+			break;
+		}
+
+		// Add ocelots in jungle biomes
+		case biJungle:
+		case biJungleHills:
+		case biJungleEdge:
+		case biJungleM:
+		case biJungleEdgeM:
+		{
+			ListOfSpawnables.insert(mtOcelot);
+			break;
+		}
+
+		// Add horses in plains-like biomes
+		case biPlains:
+		case biSunflowerPlains:
+		case biSavanna:
+		case biSavannaPlateau:
+		case biSavannaM:
+		case biSavannaPlateauM:
+		{
+			ListOfSpawnables.insert(mtHorse);
+			break;
+		}
+
+		// Add wolves in forest biomes
+		case biForest:
+		{
+			ListOfSpawnables.insert(mtWolf);
+		}
```

Comment:
A `break` seems to be missing here; if this is intentional, add a comment: `// fall-through`